src/Fx/ActionItem.cs

```
@@ -3,14 +3,19 @@
 
 namespace Microsoft.Azure.Amqp
 {
-    using System.Threading;
+    using System;
+    using System.Threading.Tasks;
 
     static class ActionItem
     {
-        public static void Schedule(WaitCallback callback, object state)
+        public static void Schedule<TState>(Action<TState> callback, TState state)
         {
             Fx.Assert(callback != null, "A null callback was passed for Schedule!");
-            ThreadPool.QueueUserWorkItem(callback, state);
+
+            _ = Task.Factory.FromAsync(
+                (stateParameter, c, callbackState) => ((Action<TState>)callbackState).BeginInvoke(stateParameter, c, callbackState),
+                r => ((Action<TState>)r.AsyncState).EndInvoke(r), state,
+                callback);
         }
```

If there is an exception from callback, the process will crash with the current code due to unhandled exception. After the change it will not happen because the exception is preserved in the task but no one is awaiting on it.

---

I wasn't aware that you want the application to crash in such a case which I find super awkward since you are essentially a library and the behavior should be up to the application to decide

In theory I could try catch on end invoke and queue the exception as a waitcallback. Thoughts?

ThreadPool.QueueUserWorkItem(new WaitCallback(ignored => 
{
   throw originalException;
}));

---

As a library it has to consider exceptions from callbacks fatal because there is no safe strategy to handle it. One example is Timer from the framework. I believe all callbacks from the library handle exceptions as appropriate. This method is also used to invoke user callbacks. In case of user callbacks not handling exceptions it should let the app crash.

---

I've added it but I'm no longer sure it is worth it. Otherwise I drop it. No hard feelings